src/main/java/io/spokestack/spokestack/nlu/NLUResult.java

```diff
@@ -0,0 +1,140 @@
+package io.spokestack.spokestack.nlu;
+
+import java.util.Map;
+
+/**
+ * A simple data class that encapsulates the result of an utterance
+ * classification.
+ */
+public final class NLUResult {
+    private final String intent;
+    private final String utterance;
+    private final Map<String, Object> context;
+    private final Map<String, Slot> slots;
+    private final Throwable error;
+
+    private NLUResult(String builderUtterance, String builderIntent,
+                      Map<String, Slot> builderSlots,
+                      Map<String, Object> builderContext) {
+        this.intent = builderIntent;
+        this.utterance = builderUtterance;
+        this.slots = builderSlots;
+        this.context = builderContext;
+        this.error = null;
+    }
+
+    private NLUResult(String builderUtterance, Throwable builderError) {
+        this.utterance = builderUtterance;
+        this.error = builderError;
+        this.intent = null;
+        this.slots = null;
+        this.context = null;
+    }
+
+    /**
+     * @return the error encountered during the NLU process.
+     */
+    public Throwable getError() {
+        return error;
+    }
+
+    /**
+     * @return The user intent.
+     */
+    public String getIntent() {
+        return intent;
+    }
+
+    /**
+     * @return The user's original utterance.
+     */
+    public String getUtterance() {
+        return utterance;
+    }
+
+    /**
+     * @return The slot values present in the user utterance.
+     */
+    public Map<String, Slot> getSlots() {
+        return slots;
+    }
+
+    /**
+     * @return Additional context included with the classification results.
+     */
+    public Map<String, Object> getContext() {
+        return context;
+    }
+
+    /**
+     * NLU result builder API.
+     */
+    public static final class Builder {
+
+        private String builderUtterance;
```

Comment:
I don't know if this is considered a best practice in java, but prefixing every variable with `builder` is silly

Reply:
The real silly thing is checkstyle complaining when your method arg has the same name as the instance field it's meant to set. That's the only reason for the dumb naming—I want the method args to have descriptive names (not really a fan of `value` from the speech pipeline builder methods, even though it's sufficient), so the silliness is hidden in private fields.